File: src/grizzle_controller/src/slider_control.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/joint_state.hpp>
#include <trajectory_msgs/msg/joint_trajectory.hpp>
#include <trajectory_msgs/msg/joint_trajectory_point.hpp>

#include <chrono>


using namespace std::chrono_literals;
using std::placeholders::_1;

class SliderControl : public rclcpp::Node
{
public:
  SliderControl() : Node("slider_control")
  {
    sub_ = create_subscription<sensor_msgs::msg::JointState>(
        "joint_commands", 10, std::bind(&SliderControl::sliderCallback, this, _1));
    left_arm_pub_ = create_publisher<trajectory_msgs::msg::JointTrajectory>("left_arm_controller/joint_trajectory", 10);
    right_arm_pub_ = create_publisher<trajectory_msgs::msg::JointTrajectory>("right_arm_controller/joint_trajectory", 10);
    RCLCPP_INFO(get_logger(), "Slider Control Node started");
  }

private:
  rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr sub_;
  rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr left_arm_pub_;
  rclcpp::Publisher<trajectory_msgs::msg::JointTrajectory>::SharedPtr right_arm_pub_;

  void sliderCallback(const sensor_msgs::msg::JointState &msg) const
  {
    trajectory_msgs::msg::JointTrajectory left_arm_command, right_arm_command;
    right_arm_command.joint_names = {"joint_1", "joint_2", "joint_3", "joint_4"};
    left_arm_command.joint_names = {"joint_5", "joint_6", "joint_7", "joint_8"};

    trajectory_msgs::msg::JointTrajectoryPoint left_arm_goal, right_arm_goal;
    left_arm_goal.positions.insert(left_arm_goal.positions.end(), msg.position.begin(), msg.position.begin() + 4);
    right_arm_goal.positions.insert(right_arm_goal.positions.end(), msg.position.begin()+4, msg.position.begin() + 8);

    left_arm_command.points.push_back(left_arm_goal);
    right_arm_command.points.push_back(right_arm_goal);

    left_arm_pub_->publish(left_arm_command);
    right_arm_pub_->publish(right_arm_command);
  }
};


int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SliderControl>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

File: src/grizzle_controller/grizzle_controller/slider_control.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import JointState
from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint


class SliderControl(Node):

    def __init__(self):
        super().__init__("slider_control")
        self.left_arm_pub_ = self.create_publisher(JointTrajectory, "left_arm_controller/joint_trajectory", 10)
        self.right_arm_pub_ = self.create_publisher(JointTrajectory, "right_arm_controller/joint_trajectory", 10)
        self.sub_ = self.create_subscription(JointState, "joint_commands", self.sliderCallback, 10)
        self.get_logger().info("Slider Control Node started")

    def sliderCallback(self, msg):
        left_arm_controller = JointTrajectory()
        right_arm_controller = JointTrajectory()
        left_arm_controller.joint_names = ["joint_1", "joint_2", "joint_3", "joint_4"]
        right_arm_controller.joint_names = ["joint_5", "joint_6", "joint_7", "joint_8"]


        left_arm_goal = JointTrajectoryPoint()
        right_arm_goal = JointTrajectoryPoint()
        left_arm_goal.positions = msg.position[:4]
        right_arm_goal.positions = msg.position[4:8]

        left_arm_controller.points.append(left_arm_goal)
        right_arm_controller.points.append(right_arm_goal)
        
        self.left_arm_pub_.publish(left_arm_controller)
        self.right_arm_pub_.publish(right_arm_controller)


def main():
    rclpy.init()

    simple_publisher = SliderControl()
    rclpy.spin(simple_publisher)
    
    simple_publisher.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
